substrate/frame/revive/src/lib.rs

```diff
@@ -544,13 +598,266 @@ pub mod pallet {
 	}
 
 	#[pallet::hooks]
-	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T>
+	where
+		// We need this to place the substrate block
+		// hash into the logs of the receipts.
+		T::Hash: frame_support::traits::IsType<H256>,
+
+		// We need these to access the ETH block gas limit via `Self::evm_block_gas_limit()`.
+		<T as frame_system::Config>::RuntimeCall:
+			Dispatchable<Info = frame_support::dispatch::DispatchInfo>,
+		T: pallet_transaction_payment::Config,
+		OnChargeTransactionBalanceOf<T>: Into<BalanceOf<T>>,
+		BalanceOf<T>: Into<U256> + TryFrom<U256>,
+		MomentOf<T>: Into<U256>,
+	{
 		fn on_idle(_block: BlockNumberFor<T>, limit: Weight) -> Weight {
 			let mut meter = WeightMeter::with_limit(limit);
 			ContractInfo::<T>::process_deletion_queue_batch(&mut meter);
 			meter.consumed()
 		}
 
+		fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
+			// Previous state root is needed to calculate the ETH block hash.
+			// TODO: Other on_initialize hooks might alter the state.
+			let version = T::Version::get().state_version();
+			let state_root = H256::decode(&mut &sp_io::storage::root(version)[..])
+				.expect("Node is configured to use the same hash; qed");
+
+			let (mut header, tx_and_receipts) = LastBlockDetails::<T>::get();
+			LastBlockDetails::<T>::kill();
+
+			header.state_root = state_root;
```

Comment:
Here we are overriding the `state_root`, which was captured in `on_finalize` of the previous block (line [824](https://github.com/paritytech/polkadot-sdk/pull/9418/files#diff-f55a8416f5e8564cb4463b5f11dfd00d7c865aabab747a05887feeedd4e1c81dR824)).
Just want to make sure I get the idea. When eg. `eth_getBlockByNumber` will be requested then `PreviousBlock` and `PreviousReceiptInfo` will be used?
If yes, then what happens if a request to fetch `latest` block is sent after `on_finalize` (block `n -1`) and before `on_initialize` (block `n`)?
In theory the latest block is `n - 1` (it was finalized), but we cannot return it before `state_root` is updated in `on_initialize` of block `n`.
But maybe I'm missing something.